### PR TITLE
Bug 1856986 - Translations UI Integration Preference Logic - Page Settings

### PR DESCRIPTION
**Pull request by iorgamgabriel**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1856986

### CODE REVIEW

**Review by ohall-m — COMMENTED** [review]
Thanks for working on this patch! Just did a quick first pass-through - it is looking good!

Question: 
The toggles weren't responding in-real time during testing, is this a planned follow-up bug or do you plan on adjusting here? (e.g., I can click the toggle, go back, and see the state change. However, the toggle doesn't immediately react when pressing.) The first toggle also seems stuck in the un-pressable state. A follow-up works for me, we just need to file and document the adjustment.

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationSwitchItem.kt [hunk 1]
```diff
@@ -4,24 +4,89 @@
 
 package org.mozilla.fenix.translations
 
+import org.mozilla.fenix.R
+
 /**
  * TranslationSwitchItem that will appear on Translation screens.
- *
+ * @property type TranslationSettingsOption type depending on the screen.
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationSwitchItem.kt:11]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationSwitchItem.kt [hunk 2]
```diff
@@ -4,24 +4,89 @@
...
+        override val hasDivider: Boolean = false,
+        override val descriptionID: Int = R.string.translation_option_bottom_sheet_switch_description,
+    ) : TranslationPageSettingsOption(hasDivider = hasDivider)
+
+    /**
+     * The page's never translate language setting.
+     */
+    data class NeverTranslateLanguage(
+        override val hasDivider: Boolean = true,
+        override val descriptionID: Int = R.string.translation_option_bottom_sheet_switch_description,
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationSwitchItem.kt:54]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationSwitchItem.kt [hunk 3]
```diff
@@ -4,24 +4,89 @@
...
+     */
+    data class AlwaysTranslateLanguage(
+        override val hasDivider: Boolean = false,
+        override val descriptionID: Int = R.string.translation_option_bottom_sheet_switch_description,
+    ) : TranslationPageSettingsOption(hasDivider = hasDivider)
+
+    /**
+     * The page's never translate language setting.
+     */
+    data class NeverTranslateLanguage(
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationSwitchItem.kt:52]: I added @params at TranslationSettingsOption; the other options only override those properties.

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsBottomSheet.kt [hunk 4]
```diff
@@ -154,13 +158,100 @@ internal fun TranslationsDialog(
 
 @Composable
 internal fun TranslationsOptionsDialog(
+    translationPageSettings: TranslationPageSettings? = null,
+    initialFrom: Language? = null,
+    context: Context,
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsBottomSheet.kt:163]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt [hunk 5]
```diff
@@ -65,23 +66,29 @@ sealed class TranslationsDialogAction : Action {
...
     /**
-     * When FetchSupportedLanguages is dispatched, an [TranslationsAction.OperationRequestedAction]
+     * When FetchSupportedLanguages is dispatched, an [TranslationOperation.FETCH_SUPPORTED_LANGUAGES]
      * will be dispatched to the [BrowserStore]
      */
-    object FetchSupportedLanguages : TranslationsDialogAction()
+    data object FetchSupportedLanguages : TranslationsDialogAction()
+
+    /**
+     * When FetchPageSettings is dispatched, an [TranslationOperation.FETCH_PAGE_SETTINGS]
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt:78]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt [hunk 6]
```diff
@@ -172,12 +187,12 @@ sealed class DismissDialogState {
...
      * The dialog should be dismissed.
      */
-    object Dismiss : DismissDialogState()
+    data object Dismiss : DismissDialogState()
 
     /**
      * This is the step when translation is in progress and the dialog is waiting to be dismissed.
      */
-    object WaitingToBeDismissed : DismissDialogState()
+    data object WaitingToBeDismissed : DismissDialogState()
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt:195]: It was a lint in android studio. 
https://stackoverflow.com/questions/76643924/object-and-data-object-in-kotlin

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogMiddleware.kt [hunk 7]
```diff
@@ -54,6 +64,42 @@ class TranslationsDialogMiddleware(
                 browserStore.dispatch(TranslationsAction.TranslateRestoreAction(sessionId))
             }
 
+            is TranslationsDialogAction.UpdateTranslationPageSettingsValue -> {
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogMiddleware.kt:67]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsBottomSheet.kt [hunk 8]
```diff
@@ -154,13 +158,100 @@ internal fun TranslationsDialog(
...
+private fun getTranslationSwitchItemList(
+    translationPageSettings: TranslationPageSettings? = null,
+    initialFrom: Language? = null,
+    context: Context,
+    onStateChange: (TranslationSettingsOption, Boolean) -> Unit,
+): List<TranslationSwitchItem> {
+    val translationSwitchItemList = mutableListOf<TranslationSwitchItem>()
+
+    translationPageSettings?.let {
+        val alwaysOfferPopup = translationPageSettings.alwaysOfferPopup
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsBottomSheet.kt:192]: Thanks for catching [1882151](https://bugzilla.mozilla.org/show_bug.cgi?id=1882151), please let me know if you spot other odd API behavior!

[the remark this answers is not in the document]
Comment (ohall-m) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsBottomSheet.kt:192]: Added [1882405](https://bugzilla.mozilla.org/show_bug.cgi?id=1882405) to fix the reducer issue.

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (ohall-m) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt:197]: Interesting, ty!

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogMiddleware.kt [hunk 9]
```diff
@@ -33,6 +34,15 @@ class TranslationsDialogMiddleware(
                 )
             }
 
+            is TranslationsDialogAction.FetchPageSettings -> {
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogMiddleware.kt:37]: TranslationsDialogAction.FetchPageSettings -> browserStore.dispatch(
                    TranslationsAction.OperationRequestedAction(
                        tabId = sessionId,
                        operation = TranslationOperation.FETCH_SUPPORTED_LANGUAGES,
                    ),
                )

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsBottomSheet.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsBottomSheet.kt:192]: Now it is working like it should. Thank you @ohall-m

File: fenix/app/src/test/java/org/mozilla/fenix/translations/TranslationsDialogMiddlewareTest.kt [hunk 10]
```diff
@@ -98,4 +100,153 @@ class TranslationsDialogMiddlewareTest {
...
+                    TranslationsAction.OperationRequestedAction(
+                        tabId = "tab1",
+                        operation = TranslationOperation.FETCH_PAGE_SETTINGS,
+                    ),
+                )
+            }
+        }
+
+    @Test
+    fun `GIVEN translationState WHEN UpdateTranslationPageSettingsValue with action type AlwaysOfferPopup from TranslationDialogStore is called THEN call UpdatePageSettingAction from BrowserStore`() =
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/test/java/org/mozilla/fenix/translations/TranslationsDialogMiddlewareTest.kt:130]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationSwitchItem.kt [hunk 11]
```diff
@@ -4,24 +4,91 @@
...
+     */
+    data class NeverTranslateSite(
+        override val hasDivider: Boolean = true,
+        override val descriptionId:
+        Int = R.string.translation_option_bottom_sheet_switch_never_translate_site_description,
+    ) : TranslationPageSettingsOption(hasDivider = hasDivider)
+}
+
+/**
+ * @property hasDivider Whether a divider should appear under the switch item.
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationSwitchItem.kt:70]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationSwitchItem.kt [hunk 12]
```diff
@@ -4,24 +4,91 @@
...
-    val importance: Int = 0,
     var isChecked: Boolean,
     var isEnabled: Boolean = true,
-    val hasDivider: Boolean,
-    val onStateChange: (Boolean) -> Unit,
+    val onStateChange: (TranslationSettingsOption, Boolean) -> Unit,
+)
+
+/**
+ * Translation settings that is related to the web-page.It will appear in [TranslationsOptionsDialog].
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationSwitchItem.kt:29]: Done

File: fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt [hunk 13]
```diff
@@ -25,7 +25,12 @@ class TranslationsDialogStore(
     initialState,
     TranslationsDialogReducer::reduce,
     middlewares,
-)
+) {
+    init {
+        dispatch(TranslationsDialogAction.FetchSupportedLanguages)
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogStore.kt:30]: Done

**Pull request merged**